File: src/ai-select/view-assessment.ts
import type { AIViewParticipation } from './ai-view';

export const aiSelectViewAssessmentPolicyVersion = 'local-view-assessment/v2';

export type ViewAssessmentStatus = 'good' | 'review' | 'failed';

/**
 * The Final Spec v1.3 §14.1 Mask Review reason vocabulary. Tracker
 * propagation and Gaussian visibility/support are not Mask-quality inputs:
 * `propagation-uncertain` is deleted (no tracker propagation exists) and
 * `weak-gaussian-support` belongs to Ticket 13 Lift Readiness.
 */
export type ReviewReason =
    | 'prompt-inconsistent'
    | 'target-materially-clipped'
    | 'severely-fragmented'
    | 'box-spill-or-neighbour-leak'
    | 'empty-or-degenerate-mask';

export type ReviewActionKey =
    | 'ai-select.review.action.inspect-mask'
    | 'ai-select.review.action.add-view'
    | 'ai-select.review.action.brush'
    | 'ai-select.review.action.inspect-view';

const actionKeysByReason: Readonly<
    Record<ReviewReason, readonly ReviewActionKey[]>
> = Object.freeze({
    'prompt-inconsistent': Object.freeze<ReviewActionKey[]>([
        'ai-select.review.action.inspect-mask',
        'ai-select.review.action.brush'
    ]),
    'target-materially-clipped': Object.freeze<ReviewActionKey[]>([
        'ai-select.review.action.inspect-view',
        'ai-select.review.action.add-view'
    ]),
    'severely-fragmented': Object.freeze<ReviewActionKey[]>([
        'ai-select.review.action.inspect-mask',
        'ai-select.review.action.brush'
    ]),
    'box-spill-or-neighbour-leak': Object.freeze<ReviewActionKey[]>([
        'ai-select.review.action.inspect-mask',
        'ai-select.review.action.brush'
    ]),
    'empty-or-degenerate-mask': Object.freeze<ReviewActionKey[]>([
        'ai-select.review.action.inspect-mask'
    ])
});

export const reviewReasonActionKeys = (
    reason: ReviewReason
): readonly ReviewActionKey[] => {
    return actionKeysByReason[reason];
};

export interface ViewAssessmentInputIdentity {
    readonly rgbDigest: string;
    readonly stableMaskDigest: string;
    readonly assessmentPolicyVersion: string;
}

/**
 * The frozen `local-view-assessment/v2` geometry thresholds, mirrored from
 * the Companion Mask Review policy so the trust boundary can reject an
 * artifact whose stated reasons are not backed by its own measured
 * diagnostics. They are version-owned: changing one requires a new policy
 * version on both sides.
 */
export const aiSelectMaskReviewThresholds = Object.freeze({
    minForegroundPixels: 4,
    fullFrameRatio: 0.98,
    clippedMinBoundaryPixels: 8,
    clippedMinBoundaryRatio: 0.2,
    fragmentMinDisconnectedPixels: 16,
    fragmentMinDisconnectedRatio: 0.1,
    boxSpillMinPixels: 16,
    boxSpillMinRatio: 0.2
});

/**
 * The measured geometry backing each structured reason. Prompt/Box
 * diagnostics are null when that Prompt family does not exist; a missing
 * family never fabricates a reason.
 */
export interface ViewAssessmentDiagnostics {
    readonly framePixels: number;
    readonly foregroundPixels: number;
    readonly boundaryPixels: number;
    readonly boundaryContactRatio: number;
    readonly connectedComponents: number;
    readonly largestComponentRatio: number;
    readonly promptPointCount: number | null;
    readonly promptViolationCount: number | null;
    readonly boxSpillPixels: number | null;
    readonly boxSpillRatio: number | null;
}

export interface ViewAssessmentResult {
    readonly status: ViewAssessmentStatus;
    readonly primaryReason?: ReviewReason;
    readonly reasons: readonly ReviewReason[];
    readonly actionableReasons: readonly ReviewReason[];
    readonly policyVersion: typeof aiSelectViewAssessmentPolicyVersion;
    readonly inputIdentity: ViewAssessmentInputIdentity;
    /**
     * Retained only for trust-boundary verification and Advanced Diagnostics.
     * Ordinary UI maps structured reason codes to static localized actions.
     */
    readonly diagnostics?: ViewAssessmentDiagnostics;
}

/**
 * The source authority behind one Stable Mask. User Confirmed authority is
 * recorded by the Mask registry and cannot be silently revoked by a later
 * automatic review.
 */
export type MaskStableAuthority = 'automatic' | 'user-confirmed';

/**
 * Final Spec v1.3 §14.2 Participation defaults, centralized here and
 * independent from View role and source: automatic Good defaults Included;
 * automatic Review, Failed, or unavailable review defaults Excluded; User
 * Confirmed defaults Included unless the user explicitly excludes (an
 * explicit controller action, never a default flip).
 */
export const defaultViewParticipation = (input: {
    readonly reviewStatus: ViewAssessmentStatus | null;
    readonly authority: MaskStableAuthority;
}): AIViewParticipation => {
    if (input.authority === 'user-confirmed') {
        return 'included';
    }
    return input.reviewStatus === 'good' ? 'included' : 'excluded';
};

type UnknownRecord = Record<string, unknown>;

const reviewReasons = new Set<ReviewReason>([
    'prompt-inconsistent',
    'target-materially-clipped',
    'severely-fragmented',
    'box-spill-or-neighbour-leak',
    'empty-or-degenerate-mask'
]);

const isRecord = (value: unknown): value is UnknownRecord => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};

const isDigest = (value: unknown): value is string => {
    return typeof value === 'string' && /^sha256:[a-f0-9]{64}$/i.test(value);
};

const isNonNegativeSafeInteger = (value: unknown): value is number => {
    return (
        typeof value === 'number' && Number.isSafeInteger(value) && value >= 0
    );
};

const isUnitInterval = (value: unknown): value is number => {
    return (
        typeof value === 'number' &&
        Number.isFinite(value) &&
        value >= 0 &&
        value <= 1
    );
};

const isNullableNonNegativeSafeInteger = (
    value: unknown
): value is number | null => {
    return value === null || isNonNegativeSafeInteger(value);
};

const isNullableUnitInterval = (value: unknown): value is number | null => {
    return value === null || isUnitInterval(value);
};

const isReviewReason = (value: unknown): value is ReviewReason => {
    return (
        typeof value === 'string' && reviewReasons.has(value as ReviewReason)
    );
};

const isReviewReasonArray = (
    value: unknown
): value is readonly ReviewReason[] => {
    return (
        Array.isArray(value) &&
        value.every(isReviewReason) &&
        new Set(value).size === value.length
    );
};

const isInputIdentity = (
    value: unknown
): value is ViewAssessmentInputIdentity => {
    return (
        isRecord(value) &&
        isDigest(value.rgbDigest) &&
        isDigest(value.stableMaskDigest) &&
        value.assessmentPolicyVersion === aiSelectViewAssessmentPolicyVersion
    );
};

const isDiagnostics = (value: unknown): value is ViewAssessmentDiagnostics => {
    return (
        isRecord(value) &&
        isNonNegativeSafeInteger(value.framePixels) &&
        isNonNegativeSafeInteger(value.foregroundPixels) &&
        isNonNegativeSafeInteger(value.boundaryPixels) &&
        isUnitInterval(value.boundaryContactRatio) &&
        isNonNegativeSafeInteger(value.connectedComponents) &&
        isUnitInterval(value.largestComponentRatio) &&
        isNullableNonNegativeSafeInteger(value.promptPointCount) &&
        isNullableNonNegativeSafeInteger(value.promptViolationCount) &&
        isNullableNonNegativeSafeInteger(value.boxSpillPixels) &&
        isNullableUnitInterval(value.boxSpillRatio)
    );
};

/**
 * Reject a fabricated claim fail-closed: every structured reason must be
 * backed by the measured diagnostic that the version-owned thresholds could
 * produce it from.
 */
const reasonsAreEvidenceBacked = (
    reasons: readonly ReviewReason[],
    diagnostics: ViewAssessmentDiagnostics
): boolean => {
    const thresholds = aiSelectMaskReviewThresholds;
    return reasons.every((reason) => {
        switch (reason) {
            case 'prompt-inconsistent':
                return (
                    diagnostics.promptViolationCount !== null &&
                    diagnostics.promptViolationCount > 0
                );
            case 'target-materially-clipped':
                return (
                    diagnostics.boundaryPixels >=
                        thresholds.clippedMinBoundaryPixels &&
                    diagnostics.boundaryContactRatio >=
                        thresholds.clippedMinBoundaryRatio
                );
            case 'severely-fragmented': {
                // The stored ratio round-trips to the exact disconnected
                // pixel mass the Companion measured.
                const disconnected =
                    diagnostics.foregroundPixels -
                    Math.round(
                        diagnostics.largestComponentRatio *
                            diagnostics.foregroundPixels
                    );
                return (
                    diagnostics.connectedComponents > 1 &&
                    disconnected >= thresholds.fragmentMinDisconnectedPixels &&
                    disconnected >=
                        thresholds.fragmentMinDisconnectedRatio *
                            diagnostics.foregroundPixels
                );
            }
            case 'box-spill-or-neighbour-leak':
                return (
                    diagnostics.boxSpillPixels !== null &&
                    diagnostics.boxSpillPixels >=
                        thresholds.boxSpillMinPixels &&
                    diagnostics.boxSpillRatio !== null &&
                    diagnostics.boxSpillRatio >= thresholds.boxSpillMinRatio
                );
            case 'empty-or-degenerate-mask':
                return (
                    diagnostics.foregroundPixels <
                        thresholds.minForegroundPixels ||
                    diagnostics.foregroundPixels >=
                        thresholds.fullFrameRatio * diagnostics.framePixels
                );
            default:
                // Unreachable over the ReviewReason union; keeps the
                // callback total for the linter.
                return false;
        }
    });
};

/**
 * Structural parse of the optional diagnostics block: `undefined` when the
 * artifact carries none, `null` when it carries a malformed one.
 */
const parseAssessmentDiagnostics = (
    value: unknown
): ViewAssessmentDiagnostics | undefined | null => {
    if (value === undefined) {
        return undefined;
    }
    return isDiagnostics(value) ? value : null;
};

export const isViewAssessmentResult = (
    value: unknown
): value is ViewAssessmentResult => {
    if (!isRecord(value)) {
        return false;
    }
    const reasons = value.reasons;
    const actionableReasons = value.actionableReasons;
    if (
        (value.status !== 'good' &&
            value.status !== 'review' &&
            value.status !== 'failed') ||
        !isReviewReasonArray(reasons) ||
        !isReviewReasonArray(actionableReasons) ||
        actionableReasons.length > 2 ||
        value.policyVersion !== aiSelectViewAssessmentPolicyVersion ||
        !isInputIdentity(value.inputIdentity)
    ) {
        return false;
    }
    const diagnostics = parseAssessmentDiagnostics(value.diagnostics);
    if (diagnostics === null) {
        return false;
    }
    if (!actionableReasons.every((reason) => reasons.includes(reason))) {
        return false;
    }
    if (value.status === 'review') {
        // Review reasons are Mask-quality claims only; the failure reason
        // never appears outside a Failed assessment.
        return (
            reasons.length > 0 &&
            !reasons.includes('empty-or-degenerate-mask') &&
            isReviewReason(value.primaryReason) &&
            value.primaryReason === reasons[0] &&
            diagnostics !== undefined &&
            reasonsAreEvidenceBacked(reasons, diagnostics)
        );
    }
    if (value.status === 'failed') {
        if (diagnostics === undefined) {
            // Assessment-internal failure: no invented reason, no geometry.
            return (
                reasons.length === 0 &&
                actionableReasons.length === 0 &&
                value.primaryReason === undefined
            );
        }
        return (
            reasons.length === 1 &&
            reasons[0] === 'empty-or-degenerate-mask' &&
            actionableReasons.length === 0 &&
            value.primaryReason === 'empty-or-degenerate-mask' &&
            reasonsAreEvidenceBacked(reasons, diagnostics)
        );
    }
    return (
        reasons.length === 0 &&
        actionableReasons.length === 0 &&
        value.primaryReason === undefined &&
        diagnostics !== undefined
    );
};

File: selection-service-companion/src/selection_service_companion/view_assessment.py
"""Versioned local per-View Mask Review for AI Select (Final Spec v1.3 §14).

The policy consumes only the exact current Mask geometry and, when the Prompt
family exists, the instance Prompt that produced the Mask. It never consumes
tracker propagation or Gaussian visibility/support: ``propagation-uncertain``
is deleted because no tracker propagation exists, and ``weak-gaussian-support``
belongs to Ticket 13 Lift Readiness, not Mask quality. Missing optional
diagnostics never fabricate a reason, and no unified confidence score exists.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


AI_SELECT_VIEW_ASSESSMENT_POLICY_VERSION = "local-view-assessment/v2"

# A Mask below this many foreground pixels cannot support a meaningful object
# decision; empty and degenerate Masks fail instead of entering Review.
_MIN_FOREGROUND_PIXELS = 4
# A Mask covering (nearly) the whole frame is not an object Mask.
_FULL_FRAME_RATIO = 0.98
# Boundary Review requires a meaningful contact margin and ratio, never any
# one-pixel contact (a thin object merely touching the edge stays Good).
_CLIPPED_MIN_BOUNDARY_PIXELS = 8
_CLIPPED_MIN_BOUNDARY_RATIO = 0.2
# Fragmentation requires material disconnected mass, not merely multiple tiny
# components: speckles next to one dominant component stay Good.
_FRAGMENT_MIN_DISCONNECTED_PIXELS = 16
_FRAGMENT_MIN_DISCONNECTED_RATIO = 0.1
# Box spill is measured outside the Box expanded by a small margin; only gross
# spill (neighbour-leak scale) is flagged.
_BOX_SPILL_MARGIN_PIXELS = 2
_BOX_SPILL_MIN_PIXELS = 16
_BOX_SPILL_MIN_RATIO = 0.2
_MAX_ACTIONABLE_REASONS = 2

ReviewReason = Literal[
    "prompt-inconsistent",
    "target-materially-clipped",
    "severely-fragmented",
    "box-spill-or-neighbour-leak",
    "empty-or-degenerate-mask",
]
AssessmentStatus = Literal["good", "review", "failed"]

_REASON_ORDER: tuple[ReviewReason, ...] = (
    "prompt-inconsistent",
    "target-materially-clipped",
    "severely-fragmented",
    "box-spill-or-neighbour-leak",
    "empty-or-degenerate-mask",
)


@dataclass(frozen=True)
class MaskReviewPrompt:
    """The instance Prompt family that produced the Mask, when one exists.

    Point/Box consistency is evaluated only when the corresponding family
    exists; an absent family contributes no reason.
    """

    positive_points: tuple[tuple[int, int], ...] = ()
    negative_points: tuple[tuple[int, int], ...] = ()
    box_xyxy: tuple[int, int, int, int] | None = None


@dataclass(frozen=True)
class LocalViewAssessmentDiagnostics:
    frame_pixels: int
    foreground_pixels: int
    boundary_pixels: int
    boundary_contact_ratio: float
    connected_components: int
    largest_component_ratio: float
    prompt_point_count: int | None
    prompt_violation_count: int | None
    box_spill_pixels: int | None
    box_spill_ratio: float | None


@dataclass(frozen=True)
class LocalViewAssessment:
    status: AssessmentStatus
    primary_reason: ReviewReason | None
    reasons: tuple[ReviewReason, ...]
    actionable_reasons: tuple[ReviewReason, ...]
    diagnostics: LocalViewAssessmentDiagnostics
    policy_version: str


def _is_foreground(mask: bytes, pixel: int) -> bool:
    return ((mask[pixel >> 3] >> (pixel & 7)) & 1) == 1


def _measure_mask(
    *, width: int, height: int, mask: bytes
) -> tuple[int, int, int, int]:
    visited = bytearray(width * height)
    foreground_pixels = 0
    boundary_pixels = 0
    components = 0
    largest_component = 0

    for pixel in range(width * height):
        if not _is_foreground(mask, pixel):
            continue
        foreground_pixels += 1
        x = pixel % width
        y = pixel // width
        if x == 0 or y == 0 or x == width - 1 or y == height - 1:
            boundary_pixels += 1
        if visited[pixel]:
            continue

        components += 1
        component_size = 0
        pending = [pixel]
        visited[pixel] = 1
        while pending:
            current = pending.pop()
            component_size += 1
            current_x = current % width
            current_y = current // width
            neighbours = (
                (current_x - 1, current_y),
                (current_x + 1, current_y),
                (current_x, current_y - 1),
                (current_x, current_y + 1),
            )
            for neighbour_x, neighbour_y in neighbours:
                if (
                    neighbour_x < 0
                    or neighbour_x >= width
                    or neighbour_y < 0
                    or neighbour_y >= height
                ):
                    continue
                neighbour = neighbour_y * width + neighbour_x
                if visited[neighbour] or not _is_foreground(mask, neighbour):
                    continue
                visited[neighbour] = 1
                pending.append(neighbour)
        largest_component = max(largest_component, component_size)

    return foreground_pixels, boundary_pixels, components, largest_component


def _count_box_spill(
    *,
    width: int,
    height: int,
    mask: bytes,
    box_xyxy: tuple[int, int, int, int],
) -> int:
    x0, y0, x1, y1 = box_xyxy
    left = max(0, x0 - _BOX_SPILL_MARGIN_PIXELS)
    top = max(0, y0 - _BOX_SPILL_MARGIN_PIXELS)
    right = min(width - 1, x1 + _BOX_SPILL_MARGIN_PIXELS)
    bottom = min(height - 1, y1 + _BOX_SPILL_MARGIN_PIXELS)
    spill = 0
    for y in range(height):
        for x in range(width):
            if left <= x <= right and top <= y <= bottom:
                continue
            if _is_foreground(mask, y * width + x):
                spill += 1
    return spill


def assess_local_view(
    *,
    width: int,
    height: int,
    mask: bytes,
    prompt: MaskReviewPrompt | None = None,
) -> LocalViewAssessment:
    """Review one exact Mask revision from version-bound local geometry."""

    if width <= 0 or height <= 0:
        raise ValueError("Mask Review dimensions must be positive")
    if len(mask) != (width * height + 7) // 8:
        raise ValueError("Mask Review mask does not match its dimensions")
    if prompt is not None:
        for point in (*prompt.positive_points, *prompt.negative_points):
            x, y = point
            if not (0 <= x < width and 0 <= y < height):
                raise ValueError("Mask Review Prompt point is outside the View")
        if prompt.box_xyxy is not None:
            x0, y0, x1, y1 = prompt.box_xyxy
            if not (0 <= x0 < x1 < width and 0 <= y0 < y1 < height):
                raise ValueError("Mask Review Prompt Box is invalid")

    foreground, boundary, components, largest = _measure_mask(
        width=width,
        height=height,
        mask=mask,
    )
    boundary_ratio = 0.0 if foreground == 0 else boundary / foreground
    largest_ratio = 0.0 if foreground == 0 else largest / foreground

    prompt_point_count: int | None = None
    prompt_violation_count: int | None = None
    box_spill_pixels: int | None = None
    box_spill_ratio: float | None = None
    if prompt is not None:
        violations = 0
        for x, y in prompt.positive_points:
            if not _is_foreground(mask, y * width + x):
                violations += 1
        for x, y in prompt.negative_points:
            if _is_foreground(mask, y * width + x):
                violations += 1
        prompt_point_count = len(prompt.positive_points) + len(
            prompt.negative_points
        )
        prompt_violation_count = violations
        if prompt.box_xyxy is not None:
            box_spill_pixels = _count_box_spill(
                width=width,
                height=height,
                mask=mask,
                box_xyxy=prompt.box_xyxy,
            )
            box_spill_ratio = (
                0.0 if foreground == 0 else box_spill_pixels / foreground
            )

    diagnostics = LocalViewAssessmentDiagnostics(
        frame_pixels=width * height,
        foreground_pixels=foreground,
        boundary_pixels=boundary,
        boundary_contact_ratio=boundary_ratio,
        connected_components=components,
        largest_component_ratio=largest_ratio,
        prompt_point_count=prompt_point_count,
        prompt_violation_count=prompt_violation_count,
        box_spill_pixels=box_spill_pixels,
        box_spill_ratio=box_spill_ratio,
    )

    # Empty, degenerate, and full-frame Masks fail closed with one structured
    # reason; they never enter Review and never invent geometry semantics.
    if (
        foreground < _MIN_FOREGROUND_PIXELS
        or foreground >= _FULL_FRAME_RATIO * width * height
    ):
        return LocalViewAssessment(
            status="failed",
            primary_reason="empty-or-degenerate-mask",
            reasons=("empty-or-degenerate-mask",),
            actionable_reasons=(),
            diagnostics=diagnostics,
            policy_version=AI_SELECT_VIEW_ASSESSMENT_POLICY_VERSION,
        )

    detected: set[ReviewReason] = set()
    if prompt_violation_count:
        detected.add("prompt-inconsistent")
    if (
        boundary >= _CLIPPED_MIN_BOUNDARY_PIXELS
        and boundary_ratio >= _CLIPPED_MIN_BOUNDARY_RATIO
    ):
        detected.add("target-materially-clipped")
    disconnected = foreground - largest
    if (
        disconnected >= _FRAGMENT_MIN_DISCONNECTED_PIXELS
        and disconnected >= _FRAGMENT_MIN_DISCONNECTED_RATIO * foreground
    ):
        detected.add("severely-fragmented")
    if (
        box_spill_pixels is not None
        and box_spill_pixels >= _BOX_SPILL_MIN_PIXELS
        and box_spill_ratio is not None
        and box_spill_ratio >= _BOX_SPILL_MIN_RATIO
    ):
        detected.add("box-spill-or-neighbour-leak")

    reasons = tuple(reason for reason in _REASON_ORDER if reason in detected)
    return LocalViewAssessment(
        status="review" if reasons else "good",
        primary_reason=reasons[0] if reasons else None,
        reasons=reasons,
        actionable_reasons=reasons[:_MAX_ACTIONABLE_REASONS],
        diagnostics=diagnostics,
        policy_version=AI_SELECT_VIEW_ASSESSMENT_POLICY_VERSION,
    )


__all__ = [
    "AI_SELECT_VIEW_ASSESSMENT_POLICY_VERSION",
    "AssessmentStatus",
    "LocalViewAssessment",
    "LocalViewAssessmentDiagnostics",
    "MaskReviewPrompt",
    "ReviewReason",
    "assess_local_view",
]
